README.md
# stackGame


Changes.md
# stackGame

So the newest version of this game added a start game scene, which contains a start button, a rotating cube and some playing instructions.

The shaders were removed, the material of cubes was replaced with MeshLambertMaterial instead.

Fixed the failing sound effect keeps playing bug.


game.js
/*

Three.js + Cannon.js video tutorial explaining the source code

Youtube: https://youtu.be/hBiGFpBle7E
SkillShare: https://skl.sh/3uREVvq 

In the tutorial, we go through the source code of this game. We cover, how to set up a Three.js scene with box objects, how to add lights, how to set up the camera, how to add animation and event handlers, and finally, we add physics simulation with Cannon.js.

Comparing to the tutorial this version has some extra features: 
- autopilot mode before the game starts
- introduction and result screens
- score indicator showing the level of layers added
- you can also control the game with touch events and by pressing the space key
- you can reset the game
- the game stops once a block went over the stack
- once the game failed the last block falls down
- the game reacts to window resizing

Check out my YouTube channel for other game tutorials: https://www.youtube.com/channel/UCxhgW0Q5XLvIoXHAfQXg9oQ

*/

window.focus(); // Capture keys right away (by default focus is on editor)

let camera, scene, renderer; // ThreeJS globals
let world; // CannonJs world
let lastTime; // Last timestamp of animation
let stack; // Parts that stay solid on top of each other
let overhangs; // Overhanging parts that fall down
const boxHeight = 1; // Height of each layer
const originalBoxSize = 3; // Original width and height of a box
let autopilot;
let gameEnded;
let robotPrecision; // Determines how precise the game is on autopilot

const scoreElement = document.getElementById('score');
const instructionsElement = document.getElementById('instructions');
const resultsElement = document.getElementById('results');

init();

// Determines how precise the game is on autopilot
function setRobotPrecision() {
  robotPrecision = Math.random() * 1 - 0.5;
}

function init() {
  autopilot = true;
  gameEnded = false;
  lastTime = 0;
  stack = [];
  overhangs = [];
  setRobotPrecision();

  // Initialize CannonJS
  world = new CANNON.World();
  world.gravity.set(0, -10, 0); // Gravity pulls things down
  world.broadphase = new CANNON.NaiveBroadphase();
  world.solver.iterations = 40;

  // Initialize ThreeJs
  const aspect = window.innerWidth / window.innerHeight;
  const width = 10;
  const height = width / aspect;

  camera = new THREE.OrthographicCamera(
    width / -2, // left
    width / 2, // right
    height / 2, // top
    height / -2, // bottom
    0, // near plane
    100 // far plane
  );

  /*
  // If you want to use perspective camera instead, uncomment these lines
  camera = new THREE.PerspectiveCamera(
    45, // field of view
    aspect, // aspect ratio
    1, // near plane
    100 // far plane
  );
  */

  camera.position.set(4, 4, 4);
  camera.lookAt(0, 0, 0);

  scene = new THREE.Scene();

  // Foundation
  addLayer(0, 0, originalBoxSize, originalBoxSize);

  // First layer
  addLayer(-10, 0, originalBoxSize, originalBoxSize, 'x');

  // Set up lights
  const ambientLight = new THREE.AmbientLight(0xffffff, 0.6);
  scene.add(ambientLight);

  const dirLight = new THREE.DirectionalLight(0xffffff, 0.6);
  dirLight.position.set(10, 20, 0);
  scene.add(dirLight);

  // Set up renderer
  renderer = new THREE.WebGLRenderer({ antialias: true });
  renderer.setSize(window.innerWidth, window.innerHeight);
  renderer.setAnimationLoop(animation);
  document.body.appendChild(renderer.domElement);
}

function startGame() {
  autopilot = false;
  gameEnded = false;
  lastTime = 0;
  stack = [];
  overhangs = [];

  if (instructionsElement) instructionsElement.style.display = 'none';
  if (resultsElement) resultsElement.style.display = 'none';
  if (scoreElement) scoreElement.innerText = 0;

  if (world) {
    // Remove every object from world
    while (world.bodies.length > 0) {
      world.remove(world.bodies[0]);
    }
  }

  if (scene) {
    // Remove every Mesh from the scene
    while (scene.children.find((c) => c.type == 'Mesh')) {
      const mesh = scene.children.find((c) => c.type == 'Mesh');
      scene.remove(mesh);
    }

    // Foundation
    addLayer(0, 0, originalBoxSize, originalBoxSize);

    // First layer
    addLayer(-10, 0, originalBoxSize, originalBoxSize, 'x');
  }

  if (camera) {
    // Reset camera positions
    camera.position.set(4, 4, 4);
    camera.lookAt(0, 0, 0);
  }
}

function addLayer(x, z, width, depth, direction) {
  const y = boxHeight * stack.length; // Add the new box one layer higher
  const layer = generateBox(x, y, z, width, depth, false);
  layer.direction = direction;
  stack.push(layer);
}

function addOverhang(x, z, width, depth) {
  const y = boxHeight * (stack.length - 1); // Add the new box one the same layer
  const overhang = generateBox(x, y, z, width, depth, true);
  overhangs.push(overhang);
}

function generateBox(x, y, z, width, depth, falls) {
  // ThreeJS
  const geometry = new THREE.BoxGeometry(width, boxHeight, depth);
  const color = new THREE.Color(`hsl(${30 + stack.length * 4}, 100%, 50%)`);
  const material = new THREE.MeshLambertMaterial({ color });
  const mesh = new THREE.Mesh(geometry, material);
  mesh.position.set(x, y, z);
  scene.add(mesh);

  // CannonJS
  const shape = new CANNON.Box(
    new CANNON.Vec3(width / 2, boxHeight / 2, depth / 2)
  );
  let mass = falls ? 5 : 0; // If it shouldn't fall then setting the mass to zero will keep it stationary
  mass *= width / originalBoxSize; // Reduce mass proportionately by size
  mass *= depth / originalBoxSize; // Reduce mass proportionately by size
  const body = new CANNON.Body({ mass, shape });
  body.position.set(x, y, z);
  world.addBody(body);

  return {
    threejs: mesh,
    cannonjs: body,
    width,
    depth,
  };
}

function cutBox(topLayer, overlap, size, delta) {
  const direction = topLayer.direction;
  const newWidth = direction == 'x' ? overlap : topLayer.width;
  const newDepth = direction == 'z' ? overlap : topLayer.depth;

  // Update metadata
  topLayer.width = newWidth;
  topLayer.depth = newDepth;

  // Update ThreeJS model
  topLayer.threejs.scale[direction] = overlap / size;
  topLayer.threejs.position[direction] -= delta / 2;

  // Update CannonJS model
  topLayer.cannonjs.position[direction] -= delta / 2;

  // Replace shape to a smaller one (in CannonJS you can't simply just scale a shape)
  const shape = new CANNON.Box(
    new CANNON.Vec3(newWidth / 2, boxHeight / 2, newDepth / 2)
  );
  topLayer.cannonjs.shapes = [];
  topLayer.cannonjs.addShape(shape);
}

window.addEventListener('mousedown', eventHandler);
window.addEventListener('touchstart', eventHandler);
window.addEventListener('keydown', function (event) {
  if (event.key == ' ') {
    event.preventDefault();
    eventHandler();
    return;
  }
  if (event.key == 'R' || event.key == 'r') {
    event.preventDefault();
    startGame();
    return;
  }
});

function eventHandler() {
  if (autopilot) startGame();
  else splitBlockAndAddNextOneIfOverlaps();
}

function splitBlockAndAddNextOneIfOverlaps() {
  if (gameEnded) return;

  const topLayer = stack[stack.length - 1];
  const previousLayer = stack[stack.length - 2];

  const direction = topLayer.direction;

  const size = direction == 'x' ? topLayer.width : topLayer.depth;
  const delta =
    topLayer.threejs.position[direction] -
    previousLayer.threejs.position[direction];
  const overhangSize = Math.abs(delta);
  const overlap = size - overhangSize;

  if (overlap > 0) {
    cutBox(topLayer, overlap, size, delta);

    // Overhang
    const overhangShift = (overlap / 2 + overhangSize / 2) * Math.sign(delta);
    const overhangX =
      direction == 'x'
        ? topLayer.threejs.position.x + overhangShift
        : topLayer.threejs.position.x;
    const overhangZ =
      direction == 'z'
        ? topLayer.threejs.position.z + overhangShift
        : topLayer.threejs.position.z;
    const overhangWidth = direction == 'x' ? overhangSize : topLayer.width;
    const overhangDepth = direction == 'z' ? overhangSize : topLayer.depth;

    addOverhang(overhangX, overhangZ, overhangWidth, overhangDepth);

    // Next layer
    const nextX = direction == 'x' ? topLayer.threejs.position.x : -10;
    const nextZ = direction == 'z' ? topLayer.threejs.position.z : -10;
    const newWidth = topLayer.width; // New layer has the same size as the cut top layer
    const newDepth = topLayer.depth; // New layer has the same size as the cut top layer
    const nextDirection = direction == 'x' ? 'z' : 'x';

    if (scoreElement) scoreElement.innerText = stack.length - 1;
    addLayer(nextX, nextZ, newWidth, newDepth, nextDirection);
  } else {
    missedTheSpot();
  }
}

function missedTheSpot() {
  const topLayer = stack[stack.length - 1];

  // Turn to top layer into an overhang and let it fall down
  addOverhang(
    topLayer.threejs.position.x,
    topLayer.threejs.position.z,
    topLayer.width,
    topLayer.depth
  );
  world.remove(topLayer.cannonjs);
  scene.remove(topLayer.threejs);

  gameEnded = true;
  if (resultsElement && !autopilot) resultsElement.style.display = 'flex';
}

function animation(time) {
  if (lastTime) {
    const timePassed = time - lastTime;
    const speed = 0.008;

    const topLayer = stack[stack.length - 1];
    const previousLayer = stack[stack.length - 2];

    // The top level box should move if the game has not ended AND
    // it's either NOT in autopilot or it is in autopilot and the box did not yet reach the robot position
    const boxShouldMove =
      !gameEnded &&
      (!autopilot ||
        (autopilot &&
          topLayer.threejs.position[topLayer.direction] <
            previousLayer.threejs.position[topLayer.direction] +
              robotPrecision));

    if (boxShouldMove) {
      // Keep the position visible on UI and the position in the model in sync
      topLayer.threejs.position[topLayer.direction] += speed * timePassed;
      topLayer.cannonjs.position[topLayer.direction] += speed * timePassed;

      // If the box went beyond the stack then show up the fail screen
      if (topLayer.threejs.position[topLayer.direction] > 10) {
        missedTheSpot();
      }
    } else {
      // If it shouldn't move then is it because the autopilot reached the correct position?
      // Because if so then next level is coming
      if (autopilot) {
        splitBlockAndAddNextOneIfOverlaps();
        setRobotPrecision();
      }
    }

    // 4 is the initial camera height
    if (camera.position.y < boxHeight * (stack.length - 2) + 4) {
      camera.position.y += speed * timePassed;
    }

    updatePhysics(timePassed);
    renderer.render(scene, camera);
  }
  lastTime = time;
}

function updatePhysics(timePassed) {
  world.step(timePassed / 1000); // Step the physics world

  // Copy coordinates from Cannon.js to Three.js
  overhangs.forEach((element) => {
    element.threejs.position.copy(element.cannonjs.position);
    element.threejs.quaternion.copy(element.cannonjs.quaternion);
  });
}

window.addEventListener('resize', () => {
  // Adjust camera
  console.log('resize', window.innerWidth, window.innerHeight);
  const aspect = window.innerWidth / window.innerHeight;
  const width = 10;
  const height = width / aspect;

  camera.top = height / 2;
  camera.bottom = height / -2;

  // Reset renderer
  renderer.setSize(window.innerWidth, window.innerHeight);
  renderer.render(scene, camera);
});


index.js
// window.focus();

const CANNON = window.CANNON;

let world;
let camera, scene, renderer;
let stackArr;
let dropsArr;
let speed;
let isGameStarted;
let isGameOver;

const score = document.getElementById('score');
const highScore = document.getElementById('highscore');

const startButton = document.getElementById('start');

const clickAudio = new Audio('./assets/click.wav');
const gameOverAudio = new Audio('./assets/gameOver.wav');
const perfectMatchAudio = new Audio('./assets/perfectMatch.wav');

let startScreenCube;

init();

function init() {
  const aspectRatio = window.innerWidth / window.innerHeight;
  const width = 10;
  const height = width / aspectRatio;
  stackArr = [];
  dropsArr = [];
  scene = new THREE.Scene();
  camera = new THREE.OrthographicCamera(
    width / -2,
    width / 2,
    height / 2,
    height / -2,
    1,
    100
  );
  camera.position.set(4, 4, 4);
  camera.lookAt(scene.position);

  const ambientLight = new THREE.AmbientLight(0xffffff, 0.6);
  scene.add(ambientLight);

  const directionalLight = new THREE.DirectionalLight(0xffffff, 0.6);
  directionalLight.position.set(10, 20, 0);
  scene.add(directionalLight);

  world = new CANNON.World();
  world.gravity.set(0, -10, 0);
  world.broadphase = new CANNON.NaiveBroadphase();
  world.solver.iterations = 40;

  renderer = new THREE.WebGLRenderer({ antialias: false });
  renderer.setSize(window.innerWidth, window.innerHeight);
  renderer.setClearColor(0x000000);
  renderer.shadowMap.enabled = true;
  renderer.shadowMap.type = THREE.PCFSoftShadowMap;
  renderer.setAnimationLoop(animation);
  document.body.appendChild(renderer.domElement);
  startButton.addEventListener('click', startGame);

  // add start screen cube

  const geometry = new THREE.BoxGeometry(2, 2, 2);
  const material = new THREE.MeshLambertMaterial({ color: 0xffa500 });
  startScreenCube = new THREE.Mesh(geometry, material);
  startScreenCube.position.set(0, 0, 0);
  scene.add(startScreenCube);
  console.log(scene.position);
}

function startGame() {
  startButton.style.display = 'none';
  scene.remove(startScreenCube);
  startScreenCube.geometry.dispose();
  startScreenCube.material.dispose();
  renderer.setAnimationLoop(animation);
  resetGame();
}

function animation() {
  speed = 0.15;

  if (startScreenCube) {
    startScreenCube.rotation.x += 0.01;
    startScreenCube.rotation.y += 0.01;
    renderer.render(scene, camera);
  }

  if (isGameStarted) {
    const topBox = stackArr[stackArr.length - 1];
    topBox.threejsBox.position[topBox.dir] += speed + stackArr.length / 1000;

    const previousBox = stackArr[stackArr.length - 2];

    const offset =
      topBox.threejsBox.position[topBox.dir] -
      previousBox.threejsBox.position[topBox.dir];
    if (offset > 10) {
      fail();
      renderer.setAnimationLoop(null);
    }
    if (camera.position.y < stackArr.length - 2 + 4) {
      camera.position.y += speed;
    }
    updateCannonjsWorld();
    renderer.render(scene, camera);
  }
}

function addBox(x, y, z, xSize, zSize, isDropping) {
  const color = new THREE.Color(`hsl(${10 + stackArr.length * 8}, 100%, 50%)`);
  const colorArray = color.toArray();

  const geometry = new THREE.BoxGeometry(xSize, 1, zSize);
  const material = new THREE.MeshLambertMaterial({ color });

  //   const vertexShader = `
  // varying vec3 vNormal;
  // varying vec3 vViewPosition;
  // void main() {
  //     vNormal = normal;
  //     vec4 viewPosition = modelViewMatrix * vec4(position, 1.0);
  //     vViewPosition = viewPosition.xyz;
  //     gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
  // }`;

  //   const fragmentShader = `
  // varying vec3 vNormal;
  // varying vec3 vViewPosition;

  // uniform vec3 lightPosition;
  // uniform vec3 lightColor;
  // uniform float shininess;
  // uniform vec3 colorArray;
  // uniform float metalness;

  // void main() {
  //   vec3 normal = normalize(vNormal);
  //   vec3 viewDirection = normalize(-vViewPosition);
  //   vec3 lightDirection = normalize(lightPosition - vViewPosition);

  //   // Diffuse Reflection
  //   float diffuse = max(dot(normal, lightDirection), 0.0);
  //   vec3 diffuseColor = diffuse * lightColor;

  //   // Specular Reflection
  //   vec3 reflectionDirection = reflect(-lightDirection, normal);
  //   float specular = pow(max(dot(reflectionDirection, viewDirection), 0.0), shininess);
  //   vec3 specularColor = specular * lightColor;

  //   // Combine Diffuse and Specular
  //   vec3 finalColor = diffuseColor + specularColor+ colorArray;

  //   gl_FragColor = vec4(finalColor, 1.0);
  // }`;

  //   const material = new THREE.ShaderMaterial({
  //     vertexShader,
  //     fragmentShader,
  //     uniforms: {
  //       lightPosition: { value: new THREE.Vector3(10, 20, 0) },
  //       lightColor: { value: new THREE.Color(0xffffff) },
  //       shininess: { value: 1 },
  //       colorArray: { value: colorArray },
  //     },
  //   });
  const threejsBox = new THREE.Mesh(geometry, material);
  threejsBox.position.set(x, y, z);
  scene.add(threejsBox);

  const boxShape = new CANNON.Box(new CANNON.Vec3(xSize / 2, 0.5, zSize / 2));
  const mass = isDropping ? 1 + xSize * zSize : 0;
  const boxBody = new CANNON.Body({ mass, shape: boxShape });
  boxBody.position.set(x, y, z);
  world.addBody(boxBody);

  return {
    threejsBox: threejsBox,
    cannonjsBox: boxBody,
    xSize,
    zSize,
  };
}

function addLayer(x, z, xSize, zSize, dir) {
  const y = stackArr.length;
  const box = addBox(x, y, z, xSize, zSize, false);
  box.dir = dir;
  stackArr.push(box);
}

function addDrop(x, z, xSize, zSize) {
  const y = stackArr.length - 1;
  const box = addBox(x, y, z, xSize, zSize, true);
  dropsArr.push(box);
}

function updateCannonjsWorld() {
  world.step(1 / 60);
  dropsArr.forEach((e) => {
    e.threejsBox.position.copy(e.cannonjsBox.position);
    e.threejsBox.quaternion.copy(e.cannonjsBox.quaternion);
  });
}

const isMobile = getDeviceType();

if (isMobile) {
  window.addEventListener('click', (e) => {
    if (!isGameOver && isGameStarted) {
      spaceDownHandler();
    } else {
      startGame();
    }
    return;
  });
} else {
  window.addEventListener('keydown', (e) => {
    if (e.key === ' ') {
      e.preventDefault();
      spaceDownHandler();
      return;
    }
    if (e.key === 'Enter') {
      e.preventDefault();
      startGame();
      return;
    }
  });
}

function spaceDownHandler() {
  if (!isGameStarted) {
    startGame();
    return;
  } else if (!isGameOver) {
    continueGame();
    return;
  }
}

function resetGame() {
  isGameOver = false;
  isGameStarted = true;
  stackArr = [];
  dropsArr = [];
  if (score) {
    score.innerText = 0;
  }
  if (highScore) {
    highScore.innerText = localStorage.getItem('highscore') || 0;
  }

  if (world) {
    world.bodies.forEach((body) => {
      world.remove(body);
    });
  }
  if (scene) {
    for (let i = scene.children.length - 1; i >= 0; i--) {
      const obj = scene.children[i];
      if (obj.type === 'Mesh') {
        scene.remove(obj);
      }
    }
    addLayer(0, 0, 3, 3);
    addLayer(-10, 0, 3, 3, 'x');
  }
  if (camera) {
    camera.position.set(4, 4, 4);
    camera.lookAt(scene.position);
  }
}

function continueGame() {
  if (isGameOver) {
    return;
  }
  const topBox = stackArr[stackArr.length - 1];
  const previousBox = stackArr[stackArr.length - 2];

  const dir = topBox.dir;
  const size = topBox[dir === 'x' ? 'xSize' : 'zSize'];
  const delta =
    topBox.threejsBox.position[dir] - previousBox.threejsBox.position[dir];

  const dropSize = Math.abs(delta);

  const overlapSize = size - dropSize;

  //   almost perfect match
  if (overlapSize > 0.95 * size) {
    perfectMatch(topBox, previousBox);
    return;
  }

  if (overlapSize > 0) {
    spiltBox(topBox, delta, overlapSize);

    const dropOffset = (overlapSize / 2 + dropSize / 2) * Math.sign(delta);
    const dropX =
      dir === 'x'
        ? topBox.threejsBox.position.x + dropOffset
        : topBox.threejsBox.position.x;

    const dropZ =
      dir === 'z'
        ? topBox.threejsBox.position.z + dropOffset
        : topBox.threejsBox.position.z;

    const dropXSize = dir === 'x' ? dropSize : topBox.xSize;
    const dropZSize = dir === 'z' ? dropSize : topBox.zSize;

    clickAudio.play();

    addDrop(dropX, dropZ, dropXSize, dropZSize);

    const nextX = dir === 'x' ? topBox.threejsBox.position.x : -10;
    const nextZ = dir === 'z' ? topBox.threejsBox.position.z : -10;
    const nextXSize = topBox.xSize;
    const nextZSize = topBox.zSize;
    const nextDir = dir === 'x' ? 'z' : 'x';

    if (score) {
      score.innerText = parseInt(score.innerText) + 1;
    }

    addLayer(nextX, nextZ, nextXSize, nextZSize, nextDir);
  } else {
    fail();
  }
}

function perfectMatch(topBox, previousBox) {
  perfectMatchAudio.play();

  topBox.threejsBox.position[topBox.dir] =
    previousBox.threejsBox.position[topBox.dir];

  const dir = topBox.dir;

  const nextX = dir === 'x' ? topBox.threejsBox.position.x : -10;
  const nextZ = dir === 'z' ? topBox.threejsBox.position.z : -10;
  const nextXSize = topBox.xSize;
  const nextZSize = topBox.zSize;
  const nextDir = dir === 'x' ? 'z' : 'x';

  if (score) {
    score.innerText = parseInt(score.innerText) + 5;
  }

  addLayer(nextX, nextZ, nextXSize, nextZSize, nextDir);
}

function spiltBox(topBox, delta, overlapSize) {
  const dir = topBox.dir;
  const newXSize = dir === 'x' ? overlapSize : topBox.xSize;
  const newZSize = dir === 'z' ? overlapSize : topBox.zSize;

  const originalSize = topBox[dir === 'x' ? 'xSize' : 'zSize'];

  topBox.xSize = newXSize;
  topBox.zSize = newZSize;

  topBox.threejsBox.scale[dir] = overlapSize / originalSize;
  topBox.threejsBox.position[dir] -= delta / 2;

  topBox.cannonjsBox.position[dir] -= delta / 2;

  const shape = new CANNON.Box(
    new CANNON.Vec3(newXSize / 2, 0.5, newZSize / 2)
  );
  topBox.cannonjsBox.shapes = [];
  topBox.cannonjsBox.addShape(shape);
}

function fail() {
  console.log('fail');
  gameOverAudio.play();

  const topBox = stackArr[stackArr.length - 1];

  addDrop(
    topBox.threejsBox.position.x,
    topBox.threejsBox.position.z,
    topBox.xSize,
    topBox.zSize
  );
  world.remove(topBox.cannonjsBox);
  scene.remove(topBox.threejsBox);

  if (score) {
    const scoreValue = parseInt(score.innerText);
    const highScoreValue = parseInt(highScore.innerText);
    if (scoreValue > highScoreValue) {
      localStorage.setItem('highscore', scoreValue);
      highScore.innerText = scoreValue;
    }
  }
  isGameOver = true;

  return;
}

// function collapse(box) {
//   const xSize = box.xSize;
//   const zSize = box.zSize;
//   const fragments = [];

//   for (let i = 0; i < 4; i++) {
//     const geometry = new THREE.BoxGeometry(xSize / 2, 1 / 2, zSize / 2);
//     const material = new THREE.MeshLambertMaterial({ color: 0xffffff });
//     const fragment = new THREE.Mesh(geometry, material);
//     fragment.position.copy(box.threejsBox.position);
//     fragment.position.x += Math.random() * 0.4 - 0.2;
//     fragment.position.y += Math.random() * 0.4 - 0.2;
//     fragment.position.z += Math.random() * 0.4 - 0.2;
//     scene.add(fragment);
//     fragments.push(fragment);
//   }

//   fragments.forEach((fragment) => {
//     const boxShape = new CANNON.Box(
//       new CANNON.Vec3(
//         fragment.geometry.parameters.width / 2,
//         fragment.geometry.parameters.height / 2,
//         fragment.geometry.parameters.depth / 2
//       )
//     );
//     const mass = 1;
//     const boxBody = new CANNON.Body({ mass, shape: boxShape });
//     boxBody.position.copy(fragment.position);
//     world.addBody(boxBody);
//   });

//   scene.remove(box.threejsBox);
//   world.remove(box.cannonjsBox);
// }

function getDeviceType() {
  const userAgent = navigator.userAgent;

  const isMobile =
    /Mobile|Android|iPhone|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
      userAgent
    );

  return isMobile;
}
